fms/utils/serialization.py

```diff
@@ -114,25 +149,32 @@ def get_ckp_format(model_path: Union[str, Path]) -> str:
 
 def load_state_dict(
     model_path: Union[str, Path],
+    model: torch.nn.Module,
+    architecture: str,
+    source: str,
     *,
     checkpoint_format: Optional[str] = None,
     distributed_strategy: Optional[str] = None,
     checkpoint_sharding: Optional[str] = None,
     initial_device: torch.device = torch.device("cpu"),
     rank: int = 0,
     world_size: int = 1,
-) -> Mapping[str, Any]:
+) -> torch.nn.Module:
```

Comment:
hmm.... it feels like this sort of replaces get_model if it's doing (almost) everything and taking almost all parameters.

I suppose we can still get a state_dict if we need to with `load_state_dict(path, model).state_dict()` but it's a bit awkward. When doing things like state dict conversions we mightwant the actual state dict. this seems a bit awkward to me.

Reply:
also one of the return types below ins incorrect; it has a mypy ignore on it but it's actually wrong with this specified return type

Reply:
the core issue here is that for safetensors (and really any other optimized format that allows for lazy loading and sharding), we need to either combine loading the state dict and sharding, or be so inefficient memory wise that we will either have to load everything on CPU memory and then copy shards to GPU or overload the GPU memory. This function still only deals with loading a model's weights from disk, while it leaves everything else (creating the model, wrapping it, etc.) to `get_model`. In fact, if no checkpoint is provided, this should never get called.